opendevin/runtime/runtime.py

```diff
@@ -54,19 +57,35 @@ class Runtime:
 
     def __init__(
         self,
-        sandbox: Sandbox | None = None,
+        event_stream: EventStream,
         sid: str = 'default',
+        sandbox: Sandbox | None = None,
     ):
         self.sid = sid
         if sandbox is None:
             self.sandbox = create_sandbox(sid, config.sandbox_type)
         else:
             self.sandbox = sandbox
         self.browser = BrowserEnv()
+        self.event_stream = event_stream
+        self.event_stream.subscribe(EventStreamSubscriber.RUNTIME, self.on_event)
+        self._bg_task = asyncio.create_task(self._start_background_observation_loop())
+
+    def close(self):
+        self.sandbox.close()
+        self.browser.close()
+        self._bg_task.cancel()
 
     def init_sandbox_plugins(self, plugins: list[PluginRequirement]) -> None:
         self.sandbox.init_plugins(plugins)
 
+    async def on_event(self, event: Event) -> None:
+        if isinstance(event, Action):
+            observation = await self.run_action(event)
+            observation._cause = event.id  # type: ignore[attr-defined]
+            source = event.source if event.source else EventSource.AGENT
```

Comment:
When will event.source have non None value here?

https://github.com/OpenDevin/OpenDevin/pull/2532#discussion_r1648856809
CC: @enyst

Reply:
Currently, that's never.

`on_event` is only called [here](https://github.com/OpenDevin/OpenDevin/blob/6879acdfe745453d336c3d8b9c597ab256d52db2/opendevin/events/stream.py#L110) and `_source` was set just above. At its turn, that parameter is non-None.

Reply:
@rbren Added for mypy sake?

CC: @xingyaoww